### ISSUE TITLE
add a way to get Entity world events

### ISSUE THREAD

**Issue by Doraku** [Doraku/DefaultEcs#36]
Concerned events:
- `EntityCreated`
- `EntityDisposed`
- `EntityEnabled`?
- `EntityDisabled`?
- `EntityComponentAdded<T>`?
- `EntityComponentRemoved<T>`?
- `EntityComponentEnabled<T>`?
- `EntityComponentDisabled<T>`?

First idea was to handle it like #34 but it seems unnatural to pass an observer to a world constructor like for an entityset to me. Also added generic events for component notification would be impossible. Maybe for world events it would be better to use the publish/subscribe mechanism. Also current namespace observer is bad a bad >_<

**Comment by captkirk88** [comment]
The addition to ISystem was one idea but a EntityAdded/EntityRemoved event would work just as well.  It fits into the pattern you have going already with publish/subscribe and it would not upset the API at all.  From there I can make my own ISystem implementation that just subscribes for those events.

**Comment by prime31** [comment]
Hmm. This seems a little heavy handed imo. I not so sure most of these events should exist in the core code. One of the core tenants of pure ECS and data oriented design is no callbacks. Having to fire off events for every mutation of the World seems like a bit much. The only one of those events that I think makes sense is `EntityDisposed` for resource cleanup.

It is trivial for a consuming application to make their own `EntityManager` that wraps `World` and has methods like CreateEntity, AddComponent, RemoveComponent, etc. Proof ;)

<img width="304" alt="Screen Shot 2019-04-10 at 6 42 53 PM" src="https://user-images.githubusercontent.com/60377/55924942-aa44e100-5bc0-11e9-92fc-f6feac82b39e.png">

That lets the game deal with creating and firing off all the mutation events if it is required. Most wouldnt need that kind of granularity unless they are creating their own custom EntitySet or replacing some other aspect of DefaultECS.

**Comment by Doraku** [comment]
Most of those events already exist in the internal of DefaultEcs and are used to update EntitySet which is why I am still hesitating, the other part of me thinking like you that the user has no business knowing about all that stuff.
For the resource cleanup there already is the ManagedResource component which hopefully handles this in a convenient way.

**Comment by prime31** [comment]
ManagedResource unfortunately doesn’t work for all cases which is why the EntityDisposed event would be nice. ManagedResource is stuck as generic constrained types which breaks blittable and unmanaged constraint so it isn’t usable in many situations.

**Comment by Doraku** [comment]
Hum you mean because TResource need to implement IDisposable? If that's the case couldn't you have TResource as a wrapper to your real blittable/unmanaged resource?

**Comment by prime31** [comment]
When a generic is involved if it isn’t constrained via “where T : unmanaged” you can’t get the sizeof the Type and blit it.

I’m also not a fan of exposing an API that has to communicate it’s intent based on unrestrained generic Types. It’s very confusing and hard to follow for someone who jumps onto the codebase. Having a nice, simple struct that asks for a resource named clearly is way easier to follow. “struct ImageAsset” vs “ManagedResource<Image, ImageLoader>”. If you didn’t write the resource code yourself you would never know what that is and how to Type and use it ;

**Comment by Doraku** [comment]
>> If you didn’t write the resource code yourself you would never know what that is and how to Type and use it ;

Haha you are probably right. I try to see the big picture when adding features but obviously sometimes my specific use cases leaks into the framework which is why I really appreciate your inputs on the different issues, I hope it helps me reach a cleaner/better solution.

On the ManagedResource, the main point of creating it was to not be linked to the enable state of the component to perform the loading, and to not expose all the inner messages used to update the sets for users of the framework to reach the same functionality. Needless to say I probably need to work much more on the documentation to explain the usage clearly.

**Comment by prime31** [comment]
The documentation for ManagedResource is fine. It explains what it does well. It's just a very specific solution to asset loading that won't work for everyone. The double generic constraint struct is an odd API and since it isnt a class it cant be subclassed to hide the generics.

I like the *idea* of the ManagedResource stack, but the API is just a bit odd. I don't have a better idea for the API at the moment though, so I have nothing to suggest ;) My asset loading system is still in its infancy and I keep changing the design of it.

One suggestion that might work around the loading on disabled Entities would be change how disabled works. Instead of a bool, the idea of "disabled" could just be a marker struct `struct Disabled {}`. That way Entities are always Entities and EntitySets are always EntitySets and there is no extra state to track/serialize/etc. You could still query for disabled structs by just asking for `With<Disabled>()`.

**Comment by Doraku** [comment]
Internally the disabled state is already handled like that with this flag
https://github.com/Doraku/DefaultEcs/blob/e2f8b79be6bca7e03f32ba67f6cdc638bb4d6f84/source/DefaultEcs/World.cs#L24
I thought about using a specific component type at first but decided against it to ensure it would not be misused (then you would be able to Set/Remove or even Disable/Enable the Disabled component which seemed strange). Also that would have meant you would need to add Without<Disabled>() intentionally to all your set construction which also seemed like a big nono for the ease of use.
Disabled is not a default state, this is something you explicitly do so it seemed more logical to let you keep track of disabled entities to be reused later on.
Actually it may be a good idea to add a GetDisabledEntities() method on World  to get a EntitySetBuilder with the disabled filter, or a WithDisabled() method on the builder to add it, that way no need for a type and you would not be able to do unexpected things hum...

If I were to take this a little further there could be a GetRecycledEntities() for disposed entities but I feel like this is definitely going too far!

If you don't mind can you share your explicit case for resource? I have a hard time thinking of an edge case with unmanaged resource so I'm probably missing something.

It's probably a ugly hack but this is how I handle blittable/nonblittable generic type with a single api
https://github.com/Doraku/DefaultEcs/blob/e2f8b79be6bca7e03f32ba67f6cdc638bb4d6f84/source/DefaultEcs/Technical/Command/ComponentCommands.cs#L39

**Comment by prime31** [comment]
For the Disabled struct, I was just thinking it would be assumed an EntitySet would be built `Without<Disabled>` as the default. Having a `WithDisabled` on EnitySetBuilder would pretty much cover the same idea.

`GetRecycledEntities` is most probably going too far! No good use case for that one!

### Resources
I currently have two kinds: managed and unmanaged. Managed resources (textures, fonts, shaders, sounds, etc) are (so far) handled using the same basic strategy. Textures as an example:

- TextureInitSystem looks for any Entities that need a texture loaded
- TextureInitSystem loads the resource and marks the resource loaded
- TextureInitSystem creates a struct either on the Entity that requested the resource or on another shared resource Entity

All resources can live on their own Entities, and then any other Entities that need them can just reference the shared Entity holding the resource struct. The Entity itself can also just keep the resource struct on itself.

All of my resource request structs are blittable while all the resource structs themselves are non-blittable since they contain the managed resource. Why blittable types? If a "Prefab" or a "Sub-Scene" can be made of only Entities and Components that are blittable, saving/loading them becomes several orders of magnitude faster.

### Unmanaged
My unmanaged resources are also blittable using a similar strategy to how Unity does things. Physics colliders are my current project and make a good example. There are many collider types all with a similar block of data (collision filters, collider type, etc) and then custom data (radius for CircleCollider, vertices for PolygonCollider, etc). All Colliders share the same header block of data and then they all have their own custom data. That whole mess of data is tossed in unmanaged memory so that a single struct can be used and *pointer casted* to the actual type of struct it is when needed.

I realize that is really hard to understand with words, so here is a diagram of sorts ;)

HeaderStruct:
[Filter]
[IsTrigger]
[ColliderType]

ColliderStruct:
[Header] <-- All collider structs must have this!

CircleColliderStruct:
[Header] <-- All collider structs must have this!
[Radius]

PhysicsCollider:
[byte* Pointer to Collider] (still blittable!)


### Some Blittable Advantages
This also makes _Scene_ composition in an editor able to save/load really fast. Lets say a _Scene_ is just the following:

- list of Systems present
- list of all Entities and their Components

If this _Scene_ can be kept blittable you can then have the idea of "play mode" and "edit mode" and make transitioning between them lightening fast.

- currently in "edit mode" adding/removing components and changing their data
- enter "play mode":
    - all Entities and all the Components need to quickly be serialized either to disk or memory
    - all "play mode" systems are added and we can test things
- exit "play mode" and move to "edit mode"
- all the data we previously serialized now needs to be reloaded to get back to where we were before entering play mode

**Comment by Doraku** [comment]
Haha what you are doing with your colliders looks a lot what I had to do for EntityCommandRecorder. Does that mean you keep the memory pinned all time for your pointers?

If I follow you correctly this would look like this:
```
// BoxCollider needs to have Equals/GetHashCode correcly overrided

public sealed class BoxColliderResourceManager : AResourceManager<BoxCollider, PhysicsCollider>
{
    protected override PhysicsCollider Load(BoxCollider info) => new PhysicsCollider { Collider = BoxCollider.Create(ref info) };

    protected override void OnResourceLoaded(in Entity entity, string info, PhysicsCollider resource) => entity.Set(resource);
}
```

ManagedResource<BoxCollider, PhysicsCollider> should still be blittable since it only contains a BoxCollider, the second generic parameter is just here so we know what message to publish to be received by the corresponding AResourceManager but I guess you also want the PhysicsCollider to be serialised so you do not have to recreate it, moving between Edit and Play mode.

**Comment by prime31** [comment]
It is very similar to what you do with EntityCommandRecorder. I don't need to pin the memory since it is 100% unmanaged via `Marshal.AllocHGlobal`. It's really ugly C#, looks a lot like C++. I have unmanaged memory pools and have to resort to using pointers to the colliders to avoid struct copies. A struct copy when the struct is really triple the size of it's current Type squashes all the other data it is hiding away and when you cast the pointer back to it's real Type all the data is junk memory!

**Comment by Doraku** [comment]
After much thinking, I ended up going back on the Observer thing on EntitySet and just put events directly on EntitySet >_>. I was probably making things more complicated than needed.
I also added a EntityDisposed event directly on World, I am not sure it will stay or if in the contrary more events will pop up there to help linking stuff more easily (you could create your own event to call when disposing, creating, ... after all, kinda like what @prime31 did with his EntityManager), though if it doesn't slow/complexify anything it would not hurt to offer those natively.
